README.md
# Water Budget Estimation and Evaluation Project (WBEEP) Visualization
This project produces an interactive, national scale, choropleth map using Hydrologic Resource Units (HRUs) as 
data groupings. A HRU is the smallest spatial unit of the Soil and Water Assessment Tool (SWAT) 
which is widely used to relate farm management practices to effects on surface waters at the
watershed scale. The HRU is not generally defined by physically meaningful boundaries but instead
lumps similar land uses, soils, and slopes within a sub-basin based upon user-defined thresholds. 

This project plots the boundaries of HRUs and colors the interior of each HRU based on calculated 
water data availability values using JavaScript, the Vue framework, and Mapbox-gl to produce a
'zoomable' national level map.



## Project setup
Clone the project. Then run 'npm install' to pull down the required dependencies into the node modules folder 
```
npm install
```

Starting the Vue development server at this point will show a website with headers and footers but no map.

```
npm run serve
```
#### UPDATE
#### We are working to consolidate and simplify working with the map locally. So, if  you pulled recent code, your map may just work.
Now you can run the map locally from S3 without the need for local tile servers. This is still in the test phase
so it still may be of benefit to read the sections after this one that explain how to set up 
the tiles to run from a local source.

But for now, here is the easy way.

There is a file in the project called mapStyles.js (src/mapStyles/mapStyles.js). This is a configuration file for Mapbox and it tells Mapbox a bunch
of important details about how to create the map. What is important to us at the moment are two things, the 'source' and the 
'source-layer.' The 'source' is where Mapbox should look for the 'tiles.' The 'source-layer' is only important to vector
tile sources, so for something like pure GeoJson you will not need it, but since our tiles are vector we will need to make
sure our source layer is correct.

The mapStyles.js is really just a ES6 JSON object called 'style', so like all JSON objects it is made up of key/value pairs.
Of main concern is the 'sources' key. In the following example, we can see that there are two sources for our map layers.
one called 'basemap' and another called 'HRU.' Mapbox needs two bits of information to process the source correctly, the 
'type' and the location. Mapbox can process many different source types, the source we are using are vector, meaning mathematical
created lines rather than dot graphed. Our vector tiles are stored in Amazon Web Services (AWS) and we have to tell that to 
Mapbox. What is important here is to know that our 'tiles' on AWS are in 'protobuf' format (.pbf). Protobuf is one of several
tile formats that Mapbox can use. To load a '.pbf' source, we need the "tiles" key along with the resource location
in the form of an array as shown below (for more information read the 'Wait! Important to know: Local tiles and AWS 
tiles are DIFFERENT!' section). 
```
    style: {
        version: 8,
        sources: {
            basemap: {
                type: "vector",
                "tiles": ["https://d38anyyapxci3p.cloudfront.net/baseTiles_3/{z}/{x}/{y}.pbf"]             
            },
            HRU: {
                type: "vector",
                "tiles": ["https://d38anyyapxci3p.cloudfront.net/tileTemp_3/{z}/{x}/{y}.pbf"]               
            }
        },
```
If you pulled the code from Github, the locations are probably correct and your map should start up with no extra effort.
However, if you are using the S3 sources and your map is not showing, or is missing layers, it may be that 
1) we moved the tiles on S3
2) you do not have the right 'source-layer' name

So, about sources and layers. Each source can be the parent of many map layers. Remember, that in the sample above, we had
two sources. That does not mean that we have only two layers. Each source can hold the information needed to make many layers.
In our map, we do just that. And we let Mapbox know which layers to make from each source by adding that information into the
Mapbox configuration file, which, as mentioned previously, is called mapStyles.js in our project.
```
            {
                id: "HRUS Fill Colors",
                type: "fill",
                source: "HRU",
                "source-layer": "no_simp_prec5", // make sure this name is right
                "layout": {
                    "visibility": "visible"
                },
                paint: {
                    "fill-color": {
                        "property": "SoilMoisture",
                        "type": 'categorical',
                        "stops": [
                            ["","#000000"],
                            ["very low","#CC4C02"],
                            ["low", "#EDAA5F"],
                            ["average","#FED98E"],
                            ["high","#A7B9D7"],
                            ["very high","#144873"],
                        ]
                    },
                    "fill-opacity": 1
                },
                "showButton": true
            },
``` 
Above, is a sample of that shows how a map layer is defined. The detail that is important here is to note the 'source-layer'
key. This has to match the name of the 'source-layer' in the tiles, and the name is defined by the person creating the tiles.
So, it sometimes changes. If you find that a layer is not showing as expected, check that the 'source-layer' has the correct name.

How do you know the name of the 'source-layer?' Well, it is not totally straight-forward. A way that works is to us a handy
program called, Vector Inspector https://stevage.github.io/vector-inspector/ . 'Vector Inspector' will give the name of
the 'source-layer' as shown circled in the image below. 
![alt text](./markDownImages/VectorInspector.png "Image of Vector Inspector")
 
Okay the above information should get your map working. However, if for some reason you chose to go 'old school' and run the tiles from a local source, there are a few things you need to know. 
1) Get the map tiles and add them to the project
2) Check that URLs for the tiles are correct
3) Start run the tile server

#### Wait! Important to know: Local tiles and AWS tiles are DIFFERENT!
Before we go on, I need to explain something. The environment on your local machine is different from that
on AWS. There are two important differences: 

First, on AWS, the built application (the contents of the 'dist' folder) is 
stored in a Simple Storage Service (S3) 'bucket.' We have told AWS to 'serve' the contents of that bucket so that
they may be used as a website. This is possible because AWS uses a simple web server to deliver the content of the 
bucket to a web browser. On your local machine, the Vue framework provides a development server that 
delivers the content to a browser.
 
Second, since AWS is a web resource, all the content on AWS has a URL.

These two things combine to influence the way our application runs in each environment. First, as we discuss in a moment,
Mapbox-gl requires map tiles to be identified by as a web resource. That means it needs the tiles to have a URL in order to 
use them. This is fine on AWS, as all the resources have URLs, however locally this will require us to use a 
tile server to generate a URL from which Mapbox-gl can fetch the tiles. Secondly, while all the resources on
AWS have URLs, AWS provides only a simple HTTP server to get those assets. Locally, our tile server(s) will do just a
bit more. The local tile server will take a single '.mbtiles' tile file, break it apart, and serve each part as an individual tile.

Now, we could have added a tile server to our application when deployed on AWS, but we wanted to keep our application
deployment simple (and inexpensive). Application deployments in AWS can use a S3 bucket as a source, only if they are static.
And by 'static' I mean they are applications that run on a users browser with the browser doing all the processing, and there
is no need for server-side-processing. Adding a tile server would mean we had added server-side-processing and we would no 
longer be able to use S3 as our host; we would have needed to switch to a more complex and expensive Elastic Compute Cloud (EC2)
instance. 

To avoid jumping up in cost and complexity to EC2, we compensated for the AWS HTTP server's inability to clip up '.mbtiles' files
by doing that part ourselves. So, on AWS we store the map tiles not as a single '.mbtiles' file but as a folder of folders,
with each folder containing sets of '.pbf' files that represent map 'zoom' level. Since we have manually done the tile cutting,
the simple HTTP server on AWS can do the rest.

Why is this important:
- on AWS, the the map tiles are 'cut' into '.pbf' files.
- on your local machine, you will have a single '.mbtiles' file (you may have several of these, but the each represent a single map layer)
- on AWS, the simple HTTP server will deliver the tile files
- on your local machine, you will need to start a tile server for each '.mbtiles' file

### Get the map tiles and add them to the project
One way to get the tiles is take them from storage on Amazon Web Services (AWS). The most current copy should be stored
at this AWS location:
```
prod-owi-resources/resources/Application/wbeep/tiles_current
```
Unfortunately there is no way to download the contents of an entire directory from S3 using the graphic
user interface, arrgh. So to get the tile files you will have to install AWS-CLI. This is accomplished by
using 'pip.' 

```
// this is an example, it is highly recommend that you do an internet search
// for the most recent install information
sudo pip install awscli
```
Installing AWS-CLI is the easy part, now you have to configure the credentials. First, you have to get the
credentials. For obvious reasons, I can not give too much detail here, and you will have to contact a team 
member for more specific advice. However, once you have the credentials you will run . . .
```
aws configure
```
This will start a program that will ask you to enter four items:
1) AWS Access Key ID 
2) AWS Secret Access Key 
3) Default region name
4) Default output format

The answers you enter will be saved at ~/.aws/credentials on Linux, macOS, or Unix, or at C:\Users\USERNAME\.aws\credentials on Windows.
I am not sure if it best practice, but I have edited this file manually without problems. 

If you open the '.aws' directory you will find two text files: 'credentials,' and 'config.' The 'config' contains the answers to the
last two questions and the name of the profile.
```
// example of 'config'

[default] // this is the profile
region = '' // this is the location of the AWS server like 'us-west-2'
output = json // this seems fine
``` 
When opened in a text editor, the credentials file looks something like this:
```
[default] // again, this is the profile
aws_access_key_id = the key id you added
aws_secret_access_key = the key you added
```
Once that is all set up, you are good to go. The AWS-CLI will find the credentials, and you will not be prompted
to enter them when you run commands such as . . .
```
// General example to copy directory from S3 to current working directory
aws s3 cp s3://mybucket . --recursive

// Specific Example for grabbing '.mbtiles' and placing them in the current working directory
// We have been using the 'tiles' directory at the project root
aws s3 cp s3://prod-owi-resources/resources/Application/wbeep/tiles_current . --recursive 
```

#### Manual Upload of Tiles
If you ever need to manually upload tiles to AWS, this command works provided you have installed AWS-CLI
on your machine and that you have configured the credentials. 
```
// generic example
aws s3 cp <local directory> <s3 target bucket> --recursive --content-encoding 'gzip' --content-type 'application/gzip'

// specific example, transfering contents of current working directory
aws s3 cp . s3://wbeep-qa-website/tiles --recursive --content-encoding 'gzip' --content-type 'application/gzip'

// Note: using the 'content-type' application/x-protobuf also seems to work, and may be the better choice
// Here a specific example
aws s3 cp . s3://wbeep-qa-website/tiles --recursive --content-encoding 'gzip' --content-type 'application/x-protobuf'
```
If you do not configure the file compression correctly, when the application is run, Mapbox-gl will not be able to read the tiles, leaving you with
a map-less web page and a cryptic error in the console.
```
Unimplemented type: 3
```
This means that everything is working except that the compression type is not what Mapbox can use. 
The tiles will have to be reloaded to AWS with the correct content-type and content-encoding.
 
A second option, which avoids the hassle of configuring the AWS-CLI credential is to use the AWS UI to upload the files. 
Here you can drag and drop the files for uploading, just make sure to set both the 'content-encoding' and the 'content-type.'
The content-encoding will be 'gzip', and the content-type will be application/gzip. Note that there is not
a specific item in the drop down menu for these choices, but they can be added in the text box.

### Check that URLs for the tiles are correct
Locally, the tiles for the map will be served by 'tile servers.' These servers will need to know where the
tiles are stored on your machine and will produce a URL that Mapbox-gl can call. When Mapbox-gl calls,
the tile server will serve up the appropriate tiles. 

Please note that Mapbox-gl requires the map tiles to have a full URL, for example 'http://localhost:8086/data/basemap.json.'
or 'wbeep-test-website.s3-website-us-west-2.amazonaws.com.' This is works fine for web based resources but makes using a relative path impossible, which means
that Mapbox-gl will require you to have a different URL for every deployment option. The path Mapbox-gl
uses are found in the style variable of the component creating the map element, such as 'MapBox.vue.'

```
    let style = {
        version: 8,
        sources: {
            basemap: {
                type: "vector",
                url: "http://localhost:8086/data/basemap.json" // here is one URL
            },
            HRU: {
                type: "vector",
                url: "http://localhost:8085/data/new2.json", // here is another

```
In the above example there are two URLs, one for the base map and a second one for the HRUs. Notice that
these are both local URLs. These URLs may differ slightly from what is required on your local machine. The main 
change would be the file names, which in the example are 'basemap.json' and 'new2.json.' These file names
need to match the names of the tile files (except the ending will be '.json' and not '.mbtiles'), which we will talk about later.

Now, since there are two URLs, we will need two tile servers, one to produce each URL. A perfectly good choice for
a tile server is tileserver-gl-light: https://www.npmjs.com/package/tileserver-gl-light . It is available as a npm
package. Follow the instructions on the tilesserver-gl-light page and install the tile server globally (use the -g flag).

### Start run the tile server
Now that you have a tile server installed locally, you can get the map up and running in the application.
The tileserver will need two items of information, the location of the tiles, and the port address on which 
to run.

```
tileserver-gl-light basemap.mbtiles -p 8086
```
The above example works when starting the tile server from directory containing the tile file to be used. In this case, the tile file
is named 'basemap.mbtiles. That is just an example name. The exact name may be different.
 
Using port 8086 is an arbitrary choice, but since the Vue development server is usually running on port 8080 this keeps conflicts to a 
minimum.  Also, the port number must match the port listed in the URL of the 'style' variable, as 

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
- Changed to use environment variables for builds and feature toggles
- Map Layers Toggle Button
- Added Google Analytics plugin
- Changed to use grouped stream orders
- Changed footer icons to Vue components
- New Map Layers Toggle Functionality
- Water layer above neighboring counties
- Restricted user ability to zoom out past the continental United States
- Added sub-subtitle with date of last available data
- Adjusted mapStyles to have HRU's solid colors
- Legend Toggle functionality
- Min-height for map
- Added footer email
- topNavToggle icon adjustment and mobile adjustment
- Changed URLs to use new tile server
- Map Options layers dynamic toggle functionality first added
- Made Subtitle more mobile friendly
- Edited Legend modal to better fit a mobile experience
- Added vue-svg-loader package
- Added Florida and Arizona SVGS
- Created Legend Modal
- Added new Legend text and legend info button
- Added a stream orders toggle bar so that we can test out which stream orders we want
- Made subtitle information modal popup
- Made Map Subtitle component
- Added map subtitle and subtitle info button
- Cleaned up mapStyles.json
- Changed map and page title to 'National Integrated Water Availability Assessments'
- Added attribution control for OpenMapTiles layers
- Removed Mapbox street layers and replaced them with OpenMapTiles
- Zoom to bounding box of the U.S.
- Removed City Dots and HRU Outlines
- Added Flexbox fluidity to the layout
- Added Mapbox layers for testing 
- Parameterized git branch/tag in Jenkinsfile
- Refined layer toggle for appearance and mobile compatibility
- Adjusted the minimum zoom level to 3 as well as the default/initial zoom level to 3
- Added Legend
- Added hover effect to HRUs
- Changed to allow tile assets to be pulled from S3
- Added layer toggle for all layers, and added 'showButton' property to mapStyles.js to control display of buttons
- Moved the map styles to a separate mapStyles.js 
- Changed application to use vue-mapbox plugin
- Dotted State lines to show its possible.
- Added Cities to the basemap layers.
- Changed basemap creation so  we have several source layers to avoid loading in all layers multiple times and filtering.  This way we only call the layer we want when we want
- changed layer order so HRU layer is lowest and changed the opacity of that layer 
- added map controls
- Deleted second mbtiles folder
- Fixed the title bar CSS at wide widths, adjusted the page title
- Added basemap mbtiles to the mbtiles directory
- Fixed menu bar title, added more appropriate title to page
- Added appropriate changes to the style variable in the Mapbox.vue to style the map the way I chose to
- Added USWDS components, Added title, removed component not needed



src/assets/mapStyles/mapStyles.js
// Since the Mapbox gl wants the tile request to have an array containing the URL for the tiles, we need to
// grab the URL for the correct tier from the environment variables and prepackage the result as an array
const hruTileUrl = [];
    hruTileUrl.push(process.env.VUE_APP_HRU_TILE_URL);

export default {
    style: {
        version: 8,
        sources: {
            basemap: {
                type: 'vector',
                'tiles': ['https://maptiles-prod-website.s3-us-west-2.amazonaws.com/misctilesets/usstatecounties/{z}/{x}/{y}.pbf'],
                'minzoom': 2, // setting this to equal the minzoom of main map, real tile extent is 2
                'maxzoom': 10  // setting this to the real tile extent is 10
            },
            HRU: {
                type: 'vector',
                'tiles': hruTileUrl,
                'minzoom': 2, // setting this to equal the minzoom of main map, real tile extent is 0
                'maxzoom': 9  // setting this to equal the maxzoom of main map, real tile extent is 11
            },
            nhd_streams_grouped: {
                type: 'vector',
                'tiles':['http://maptiles-prod-website.s3-us-west-2.amazonaws.com/nhdstreams_grouped/{z}/{x}/{y}.pbf'],
                'minzoom': 2, // setting this to equal the minzoom of main map, real tile extent is 0
                'maxzoom': 9  // setting this to equal the maxzoom of main map, real tile extent is 10
            },
            openmaptiles: {
                type: 'vector',
                'tiles': ['https://maptiles-prod-website.s3-us-west-2.amazonaws.com/openmaptiles/baselayers/{z}/{x}/{y}.pbf'],
                'minzoom': 2,
                'maxzoom': 14
            },
            hillshade: {
                type: 'raster',
                'tiles': ['https://maptiles-prod-website.s3-us-west-2.amazonaws.com/openmaptiles/omthillshade/{z}/{x}/{y}.png'],
                'minzoom': 2,
                'maxzoom': 12,
                'tileSize': 256
            }
        },
        'sprite': '',
        'glyphs': 'https://orangemug.github.io/font-glyphs/glyphs/{fontstack}/{range}.pbf',
        'layers': [
            {
                'id': 'background',
                'paint': {
                    'background-color': 'hsl(47, 26%, 88%)'
                },
                'type': 'background',
                'showButtonLayerToggle': false
            },
            {
                'id': 'HRUs',
                'type': 'fill',
                'source': 'HRU',
                'source-layer': 'hrus',
                'layout': {
                    'visibility': 'visible'
                },
                'paint': {
                    'fill-color': {
                        'property': 'value',
                        'type': 'categorical',
                        'stops': [
                            ['very high','#144873'],
                            ['high','#A7B9D7'],
                            ['average','#FED98E'],
                            ['low', '#EDAA5F'],
                            ['very low','#CC4C02'],
                            ['','#000000'],
                        ]
                    },
                    'fill-opacity': ['case',
                        ['boolean', ['feature-state', 'hover'], false],
                        0.7,
                        1
                    ],
                    'fill-outline-color': ['case',
                        ['boolean', ['feature-state', 'hover'], false],
                        'rgba(0, 0, 0, 1)',
                        'rgba(0, 0, 0, 0.0)'
                    ],
                },
                'showButtonLayerToggle': false,
                'legendText': {
                    'very high': ['Uncommonly Wet',' \- it\'s typically drier than it is today in this region'],
                    'high': ['',''],
                    'average': ['Common',' \- today is normal for this region'],
                    'low': ['',''],
                    'very low': ['Uncommonly Dry',' \- it\'s typically wetter than it is today in this region'],
                    'no data': ['','No Data']
                }
            },
            {
                'id': 'hill shade',
                'type': 'raster',
                'source': 'hillshade',
                'layout': {
                    'visibility': 'visible'
                },
                'showButtonLayerToggle': true,
                'showButtonStreamToggle': false,
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['in', 'class', 'minor', 'service']
                ],
                'id': 'road_minor',
                'layout': {
                    'line-cap': 'round',
                    'line-join': 'round'
                },
                'paint': {
                    'line-color': 'hsl(0, 0%, 97%)',
                    'line-width': {
                        'base': 1.55,
                        'stops': [
                            [4, 0.25],
                            [20, 30]
                        ]
                    }
                },
                'source': 'openmaptiles',
                'source-layer': 'transportation',
                'type': 'line',
                'minzoom': 13,
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['in', 'class', 'trunk', 'primary']
                ],
                'id': 'road_trunk_primary',
                'layout': {
                    'line-cap': 'round',
                    'line-join': 'round'
                },
                'paint': {
                    'line-color': '#fff',
                    'line-width': {
                        'base': 1.4,
                        'stops': [
                            [6, 0.5],
                            [20, 30]
                        ]
                    }
                },
                'source': 'openmaptiles',
                'source-layer': 'transportation',
                'type': 'line',
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['in', 'class', 'secondary', 'tertiary']
                ],
                'id': 'road_secondary_tertiary',
                'layout': {
                    'line-cap': 'round',
                    'line-join': 'round'
                },
                'paint': {
                    'line-color': '#fff',
                    'line-width': {
                        'base': 1.4,
                        'stops': [
                            [6, 0.5],
                            [20, 20]
                        ]
                    }
                },
                'source': 'openmaptiles',
                'source-layer': 'transportation',
                'type': 'line',
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['==', 'class', 'motorway']
                ],
                'id': 'road_major_motorway',
                'minzoom': 5,
                'layout': {
                    'line-cap': 'round',
                    'line-join': 'round'
                },
                'paint': {
                    'line-color': 'hsl(0, 0%, 100%)',
                    'line-offset': 0,
                    'line-width': {
                        'base': 1.4,
                        'stops': [
                            [8, 1],
                            [16, 10]
                        ]
                    }
                },
                'source': 'openmaptiles',
                'source-layer': 'transportation',
                'type': 'line',
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'Polygon'],
                    ['==', 'intermittent', 1]
                ],
                'id': 'water_intermittent',
                'paint': {
                    'fill-color': 'hsl(205, 56%, 73%)',
                    'fill-opacity': 0.7
                },
                'source': 'openmaptiles',
                'source-layer': 'water',
                'type': 'fill',
                'layout': {
                    'visibility': 'visible'
                },
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['==', 'brunnel', 'tunnel']
                ],
                'id': 'waterway-tunnel',
                'paint': {
                    'line-color': 'hsl(205, 56%, 73%)',
                    'line-dasharray': [3, 3],
                    'line-gap-width': {
                        'stops': [
                            [12, 0],
                            [20, 6]
                        ]
                    },
                    'line-opacity': 1,
                    'line-width': {
                        'base': 1.4,
                        'stops': [
                            [8, 1],
                            [20, 2]
                        ]
                    }
                },
                'source': 'openmaptiles',
                'source-layer': 'waterway',
                'type': 'line',
                'layout': {
                    'visibility': 'visible'
                },
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['!in', 'brunnel', 'tunnel', 'bridge'],
                    ['!=', 'intermittent', 1]
                ],
                'id': 'waterway',
                'paint': {
                    'line-color': 'hsl(205, 56%, 73%)',
                    'line-opacity': 1,
                    'line-width': {
                        'base': 1.4,
                        'stops': [
                            [8, 1],
                            [20, 8]
                        ]
                    }
                },
                'source': 'openmaptiles',
                'source-layer': 'waterway',
                'type': 'line',
                'layout': {
                    'visibility': 'visible'
                },
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'LineString'],
                    ['!in', 'brunnel', 'tunnel', 'bridge'],
                    ['==', 'intermittent', 1]
                ],
                'id': 'waterway_intermittent',
                'paint': {
                    'line-color': 'hsl(205, 56%, 73%)',
                    'line-opacity': 1,
                    'line-width': {
                        'base': 1.4,
                        'stops': [
                            [8, 1],
                            [20, 8]
                        ]
                    },
                    'line-dasharray': [2, 1]
                },
                'source': 'openmaptiles',
                'source-layer': 'waterway',
                'type': 'line',
                'layout': {
                    'visibility': 'visible'
                },
                'showButtonLayerToggle': false
            },
            {
                'id': 'large rivers',
                'nhdOrder': 'stream_order_6',
                'type': 'line',
                'source': 'nhd_streams_grouped',
                'source-layer': 'nhdplus_order_6',
                'layout': {
                    'visibility': 'none'
                },
                'paint': {
                    'line-color': 'rgba(115, 255, 255, 1)'
                },
                'showButtonLayerToggle': false,
                'showButtonStreamToggle': true
            },
            {
                'id': 'rivers',
                'nhdOrder': 'nhdplus_orders_4_5',
                'type': 'line',
                'source': 'nhd_streams_grouped',
                'source-layer': 'nhdplus_orders_4_5',
                'layout': {
                    'visibility': 'none'
                },
                'paint': {
                    'line-color': 'rgba(115, 255, 255, 1)'
                },
                'showButtonLayerToggle': false,
                'showButtonStreamToggle': true
            },
            {
                'id': 'streams',
                'nhdOrder': 'nhdplus_orders_2_3',
                'type': 'line',
                'source': 'nhd_streams_grouped',
                'source-layer': 'nhdplus_orders_2_3',
                'layout': {
                    'visibility': 'none'
                },
                'paint': {
                    'line-color': 'rgba(115, 255, 255, 1)'
                },
                'showButtonLayerToggle': false,
                'showButtonStreamToggle': true
            },
            {
                'id': 'small tributaries',
                'nhdOrder': 'stream_order_1',
                'type': 'line',
                'source': 'nhd_streams_grouped',
                'source-layer': 'nhdplus_order_1',
                'layout': {
                    'visibility': 'none'
                },
                'paint': {
                    'line-color': 'rgba(115, 255, 255, 1)'
                },
                'showButtonLayerToggle': false,
                'showButtonStreamToggle': true,
            },

            {
                'id': 'Neighboring Countries',
                'type': 'fill',
                'source': 'basemap',
                'minzoom': 2,
                'maxzoom': 24,
                'source-layer': 'neighboringcountry',
                'layout': {
                    'visibility': 'visible'
                },
                'paint': {
                    'fill-color': 'rgba(237, 236, 232, 1)'
                },
                'showButtonLayerToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'Polygon'],
                    ['!=', 'intermittent', 1]
                ],
                'id': 'water',
                'paint': {
                    'fill-color': 'hsl(205, 56%, 73%)'
                },
                'source': 'openmaptiles',
                'source-layer': 'water',
                'type': 'fill',
                'layout': {
                    'visibility': 'visible'
                },
                'showButtonLayerToggle': false
            },
            {
                'id': 'counties',
                'type': 'line',
                'source': 'basemap',
                'source-layer': 'counties',
                'minzoom': 6,
                'maxzoom': 24,
                'layout': {
                    'visibility': 'visible'
                },
                'paint': {
                    'line-color': 'rgb(0,0,0)',
                    'line-dasharray': [4, 3]
                },
                'showButtonLayerToggle': false
            },
            {
                'id': 'states',
                'type': 'line',
                'source': 'basemap',
                'source-layer': 'states',
                'minzoom': 2,
                'maxzoom': 24,
                'layout': {
                    'visibility': 'visible',
                },
                'paint': {
                    'line-color': 'rgb(0,0,0)'
                }

            },
            {
                'id': 'Hydrological Response Unit',
                'type': 'line',
                'source': 'HRU',
                'source-layer': 'hrus',
                'layout': {
                    'visibility': 'none'
                },
                'paint': {
                    'line-color': 'rgba(57, 79, 87, 1)'
                },
                'showButtonLayerToggle': true,
                'showButtonStreamToggle': false
            },
            {
                'filter': ['all', ['==', '$type', 'Point'],
                    ['==', 'class', 'city']
                ],
                'id': 'place_label_city',
                'layout': {
                    'text-field': '{name:latin}\n{name:nonlatin}',
                    'text-font': ['Noto Sans Regular'],
                    'text-max-width': 10,
                    'text-size': {
                        'stops': [
                            [3, 12],
                            [8, 16]
                        ]
                    }
                },
                'maxzoom': 16,
                'minzoom': 5,
                'paint': {
                    'text-color': 'hsl(0, 0%, 0%)',
                    'text-halo-blur': 0,
                    'text-halo-color': 'hsla(0, 0%, 100%, 0.75)',
                    'text-halo-width': 2
                },
                'source': 'openmaptiles',
                'source-layer': 'place',
                'type': 'symbol',
                'showButtonLayerToggle': false
            }
        ]
    }
};
